pybossa/cache/projects.py

```diff
@@ -630,6 +630,23 @@ def get_from_pro_user():
         projects.append(project)
     return projects
 
+@memoize(timeout=ONE_DAY)
+def get_recently_updated_projects():
+    """Return the list of published projects that has task creations in last 3 months."""
+
+    sql = text('''SELECT t.project_id, p.short_name
+               FROM task t INNER JOIN project p
+               ON t.project_id = p.id
+               GROUP BY t.project_id, p.short_name, p.id
```

Comment:
Two minor comments:
- We might not need to group t.project_id
- Could be a good idea to select the smaller size table first.

Reply:
> * We might not need to group t.project_id
this is required, otherwise sql reports error
`ERROR:  column "t.project_id" must appear in the GROUP BY clause or be used in an aggregate function`

> * Could be a good idea to select the smaller size table first.
explain plan for both looks same
```
pybossa=# explain SELECT t.project_id, p.short_name FROM task t INNER JOIN project p ON t.project_id = p.id GROUP BY t.project_id, p.short_name, p.published HAVING p.published = true AND TO_DATE(max(t.created), 'YYYY-MM-DD"T"HH24:MI:SS.U
S') >= NOW() - '3 months'  :: INTERVAL;
                                               QUERY PLAN                                                
---------------------------------------------------------------------------------------------------------
 HashAggregate  (cost=31.35..31.47 rows=6 width=40)
   Group Key: t.project_id, p.short_name, p.published
   Filter: (to_date(max(t.created), 'YYYY-MM-DD"T"HH24:MI:SS.US'::text) >= (now() - '3 mons'::interval))
   ->  Nested Loop  (cost=0.13..31.29 rows=6 width=40)
         Join Filter: (t.project_id = p.id)
         ->  Seq Scan on task t  (cost=0.00..1.09 rows=9 width=31)
         ->  Materialize  (cost=0.13..29.94 rows=2 width=13)
               ->  Index Scan using project_pkey on project p  (cost=0.13..29.93 rows=2 width=13)
                     Filter: published
(9 rows)

Time: 1.229 ms
pybossa=# 
pybossa=# 
pybossa=# explain SELECT t.project_id, p.short_name FROM project p INNER JOIN task t ON t.project_id = p.id GROUP BY t.project_id, p.short_name, p.published HAVING p.published = true AND TO_DATE(max(t.created), 'YYYY-MM-DD"T"HH24:MI:SS.U
S') >= NOW() - '3 months'  :: INTERVAL;
                                               QUERY PLAN                                                
---------------------------------------------------------------------------------------------------------
 HashAggregate  (cost=31.35..31.47 rows=6 width=40)
   Group Key: t.project_id, p.short_name, p.published
   Filter: (to_date(max(t.created), 'YYYY-MM-DD"T"HH24:MI:SS.US'::text) >= (now() - '3 mons'::interval))
   ->  Nested Loop  (cost=0.13..31.29 rows=6 width=40)
         Join Filter: (p.id = t.project_id)
         ->  Seq Scan on task t  (cost=0.00..1.09 rows=9 width=31)
         ->  Materialize  (cost=0.13..29.94 rows=2 width=13)
               ->  Index Scan using project_pkey on project p  (cost=0.13..29.93 rows=2 width=13)
                     Filter: published
(9 rows)

```

Reply:
- try to select p.id instead of t.project_id. Getting rid of less "group by" should give you back hundreds mili seconds, although might not be significant.
-  You see the same explain plan because the SQL parser optimize for us. If you apply the first item I mentioned, it makes more sense to select from "project" table first.

Reply:
didn't observed significant difference in query execution time. however, it make sense to reduce one field from grouping. Thanks!

Reply:
updated.